test/test-viz.cpp

```diff
@@ -0,0 +1,311 @@
+                element.ins.push_back(j);
+                element.layer = scheme[j].layer + 1;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    //Method to search where scheme overlaps queue and to search for connection in those elements
+    void queue_handling() {
+        while(!queue.empty()) {
+            for(int i = 0; i < scheme.size(); i++) {
+                for(int j = 0; j < queue.size(); j++) {
+                    if(scheme[i].name == std::get<0>(queue[j]).name) {
+                        if(connection_search(scheme[i], std::get<1>(queue[j]))) {
+                            queue.erase(queue.begin() + j);
+                            j--;
+                            continue;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void output(std::ostream& out) {
+        for(Vertex& vert : scheme) {
+            out << vert.type << ' ' << vert.name << ' ' << vert.layer << " INS: ";
+            for(auto ins : vert.ins) {
+                out << scheme[ins].name << ' ';
+            }
+            std::cout << std::endl;
+        }
+        std::cout << "\n====QUEUE===\n";
+        for(auto &requests : queue) {
+            out << std::get<0>(requests).type;
+        }
+    }
+
+    std::vector<Vertex> scheme;
+    std::vector<std::tuple<Vertex, std::string>> queue;
+    int outputNum;
+};
+
+class bench_parser : public lorina::bench_reader {
+public:
+    explicit bench_parser(logic_scheme& logicScheme) : scheme(logicScheme) {}
+
+    virtual void on_input(const std::string& name) const override {
+        scheme.add_input(name);
+    }
+
+    virtual void on_output(const std::string& name) const override {
+        scheme.add_output(name);
+    }
+
+    virtual void on_dff(const std::string &input, const std::string &output) const override {
+        scheme.add_dff(input, output);
+    }
+
+    virtual void on_gate(const std::vector<std::string>& inputs, const std::string& output, const std::string& type) const override {
+        scheme.add_gate(inputs, output, type);
+    }
+
+    logic_scheme& scheme;
+};
+
+void setPosition(std::vector<Vertex*> elements, const int max_elements, const float height, const float width, const int screen_height) {
+    int count = 0;
+    int freeSpaces = elements.size() + 1;
+    float step = (screen_height - height * elements.size()) / freeSpaces;
+    for(Vertex* elem : elements) {
+        elem->position.x = elem->layer * HSPACING * width;
+        elem->position.h = height;
+        elem->position.w = width;
+
+        if (elements.size() == max_elements) {
+            elem->position.y = count * VSPACING * height;
+        }
+        else {
+            elem->position.y = step * (count + 1) + height * count;
+        }
+
+        count++;
+    }
+}
+
+void setConnections(Vertex* element, const std::vector<Vertex>& scheme, std::vector<Line>& connections) {
+    for(int& input : element->ins) {
```

Comment:
Do not iterate by non-const reference when mutation is not required: iterate by value for primitive types and by const-reference for structures.